src/null-safety/understanding-null-safety/index.md

```diff
@@ -0,0 +1,1410 @@
+you can see, it can be instantiated with either kind. The answer is that `T` a
+*potentially nullable type*. Inside the body of a generic class or method, a
+potentially nullable type has all of the restrictions of both nullable types
+*and* non-nullable types.
+
+The former means you can't call any methods on it except the handful defined on
+Object. The latter means that you must initialize any fields or variables of
+that type before they're used. This can make type parameters pretty hard to work with.
+
+In practice, a few patterns show up. In collection-like classes where the type
+parameter can be instantiated with any type at all, you just have to deal with
+the restrictions. In most cases, like the example here, it means ensuring you do
+have access to a value of the type argument's type whenever you need to work
+with one. Fortunately, collection-like classes rarely call methods on their
+elements.
+
+In places where you don't have access to a value, you can make the type
+parameter nullable:
+
+```dart
+class Box<T> {
+  final T? object;
+  Box.empty();
+  Box.full(this.object);
+}
+```
+
+Note the `?` on the declaration of `object`. Now the field has an explicitly
+nullable type, so it is fine to leave it uninitialized.
+
+When you make a type parameter type nullable like `T?` here, you may need to
+cast the nullability away. The correct way to do that is using an explicit `as
+T` cast, *not* the `!` operator:
+
+```dart
+class Box<T> {
+  final T? object;
+  Box.empty();
+  Box.full(this.object);
+
+  T unbox() => object as T;
+}
+```
+
+The `!` operator *always* throws if the value is `null`. But if the type
+parameter has been instantiated with a nullable type, then `null` is a perfectly
+valid value for `T`:
+
+```dart
+main() {
+  var box = Box<int?>.full(null);
+  print(box.unbox());
+}
+```
+
+This program should run without error. Using `as T` accomplishes that. Using
+`!` would throw an exception.
+
+Other generic types have some bound that restricts the kinds of type arguments
+that can be applied:
+
+```dart
+Rect<T extends num> {
+  final T left, top, right, bottom;
+
+  Rect(this.left, this.top, this.right, this.bottom);
+
+  T get area => (right - left) * (bottom - top);
+}
+```
+
+If the bound is non-nullable, then the type parameter is also non-nullable. This
+means you have the restrictions of non-nullable types&mdash;you can't leave
+fields and variables uninitialized. But you can call any methods on values of
+the type parameter type that are declared on its bound. Having a non-nullable
+bound does, however, prevent *users* of your generic class from instantiating it
+with a nullable type argument. That's probably a reasonable limitation for most
+classes.
+
+You can also use a nullable *bound*:
+
+```dart
+Rect<T extends num?> {
+  final T left, top, right, bottom;
+
+  T get area => (right - left) * (bottom - top);
+}
+```
+
+This means that in the body of the class you get the flexibility of treating the
```

Comment:
It's a little confusing here whether you intend the code to be invalid or not.  As written, the definition of `area` is invalid, so maybe call that out in some way?

Reply:
Oops, I just totally forgot to update the code. Replaced it with a better example.